### ISSUE TITLE
Investigate significant values on the master node

### ISSUE THREAD

**Issue by javiersanchz** [wazuh/wazuh#27058]
## Description

During the workload benchamarks metrics pipeline execution for the [Release 4.10.0 - Beta 1 - Workload benchmarks metrics](https://github.com/wazuh/wazuh/issues/26995), the cluster performance tests failed because of high `USS (KB)` and `FD` values on the master.

<details>

<summary>test_cluster_performance</summary>

```console
(qa) wazuh@javier:~/Git/wazuh-qa/wazuh-qa/tests/performance/test_cluster$ python3 -m pytest test_cluster_performance --artifacts_path='/home/wazuh/Escritorio/artifacts' --n_workers=25 --n_agents=50000 --html=report.html --self-contained-html
=============================================================================================== test session starts ===============================================================================================
platform linux -- Python 3.9.16, pytest-7.1.2, pluggy-1.0.0
rootdir: /home/wazuh/Git/wazuh-qa/wazuh-qa/tests, configfile: pytest.ini
plugins: metadata-2.0.4, testinfra-5.0.0, html-3.1.1
collected 1 item                                                                                                                                                                                                  

test_cluster_performance/test_cluster_performance.py F                                                                                                                                                      [100%]

==================================================================================================== FAILURES =====================================================================================================
____________________________________________________________________________________________ test_cluster_performance _____________________________________________________________________________________________

artifacts_path = '/home/wazuh/Escritorio/artifacts', n_workers = 25, n_agents = 50000

    def test_cluster_performance(artifacts_path, n_workers, n_agents):
        """Check that a cluster environment did not exceed certain thresholds.
    
        This test obtains various statistics (mean, max, regression coefficient) from CSVs with
        data generated in a cluster environment (resources used and duration of tasks). These
        statistics are compared with thresholds established in the data folder.
    
        Args:
            artifacts_path (str): Path where CSVs with cluster information can be found.
            n_workers (int): Number of workers folders that are expected inside the artifacts path.
            n_agents (int): Number of agents in the cluster environment.
        """
        if None in (artifacts_path, n_workers, n_agents):
            pytest.fail("Parameters '--artifacts_path=<path> --n_workers=<n_workers> --n_agents=<n_agents>' are required.")
    
        # Check if there are threshold data for the specified number of workers and agents.
        selected_conf = f"{n_workers}w_{n_agents}a"
        if selected_conf not in configurations:
            pytest.fail(f"This is not a supported configuration: {selected_conf}. "
                        f"Supported configurations are: {', '.join(configurations.keys())}.")
    
        # Check if path exists and if expected number of workers matches what is found inside artifacts.
        try:
            cluster_info = ClusterEnvInfo(artifacts_path).get_all_info()
        except FileNotFoundError:
            pytest.fail(f"Path '{artifacts_path}' could not be found or it may not follow the proper structure.")
    
        if cluster_info.get('worker_nodes', 0) != int(n_workers):
            pytest.fail(f"Information of {n_workers} workers was expected inside the artifacts folder, but "
                        f"{cluster_info.get('worker_nodes', 0)} were found.")
    
        # Calculate stats from data inside artifacts path.
        data = {'tasks': ClusterCSVTasksParser(artifacts_path).get_stats(),
                'resources': ClusterCSVResourcesParser(artifacts_path).get_stats()}
    
        if not data['tasks'] or not data['resources']:
            pytest.fail(f"Stats could not be retrieved, '{artifacts_path}' path may not exist, it is empty or it may not"
                        f" follow the proper structure.")
    
        # Compare each stat with its threshold.
        for data_name, data_stats in data.items():
            for phase, files in data_stats.items():
                for file, columns in files.items():
                    for column, nodes in columns.items():
                        for node_type, stats in nodes.items():
                            for stat, value in stats.items():
                                th_value = configurations[selected_conf][data_name][phase][file][column][node_type][stat]
                                if value[1] >= th_value:
                                    exceeded_thresholds.append({'value': value[1], 'threshold': th_value, 'stat': stat,
                                                                'column': column, 'node': value[0], 'file': file,
                                                                'phase': phase})
    
        try:
>           assert not exceeded_thresholds, 'Some thresholds were exceeded:\n- ' + '\n- '.join(
                '{stat} {column} {value} >= {threshold} ({node}, {file}, {phase})'.format(**item) for item in
                exceeded_thresholds)
E               AssertionError: Some thresholds were exceeded:
E                 - reg_cof USS(KB) 1288.8672659968279 >= 850 (master, wazuh-clusterd, stable_phase)
E                 - reg_cof FD 0.8511898466419888 >= 0.6 (master, wazuh-clusterd, stable_phase)
E               assert not [{'column': 'USS(KB)', 'file': 'wazuh-clusterd', 'node': 'master', 'phase': 'stable_phase', ...}, {'column': 'FD', 'file': 'wazuh-clusterd', 'node': 'master', 'phase': 'stable_phase', ...}]

test_cluster_performance/test_cluster_performance.py:85: AssertionError
---------------------------------------------------------------------------------------------- Captured stdout call -----------------------------------------------------------------------------------------------

Setup phase took 0:18:22s (2024/11/25 12:41:28 - 2024/11/25 12:59:50).
Stable phase took 0:06:33s (2024/11/25 12:59:50 - 2024/11/25 13:06:23).
---------------------------------------------------- generated html file: file:///home/wazuh/Git/wazuh-qa/wazuh-qa/tests/performance/test_cluster/report.html -----------------------------------------------------
============================================================================================= short test summary info =============================================================================================
FAILED test_cluster_performance/test_cluster_performance.py::test_cluster_performance - AssertionError: Some thresholds were exceeded:
================================================================================================ 1 failed in 0.92s ================================================================================================
```

</details>

The values were compared with those from the previous release: [Release 4.10.0 - Alpha 3 - Workload benchmarks metrics](https://github.com/wazuh/wazuh/issues/26811)

![FD_grafico](https://github.com/user-attachments/assets/7483fbb9-5b2a-4752-b8ea-14f6784f9e93)

![USS(KB)_grafico](https://github.com/user-attachments/assets/5119dc99-b5a9-485d-a4df-5e609653d08e)

As can be seen, the `FD` values are similar in both cases, but the `USS(KB)` graph shows a significant difference between them, which we need to investigate further.

It was also compared with the `BETA1_4.9.1-RC2` branch. The file with the graphs is attached: 
[graphics.zip](https://github.com/user-attachments/files/17920320/graphics.zip)

Artifacts : [artifacts.zip](https://github.com/user-attachments/files/17920325/artifacts.zip)

**Comment by javiersanchz** [comment]
## Update

- I have been looking for possible errors, failures... that could cause the growth of these values in the master node within the generated artifacts, but I haven't found anything new that hasn't already been reported.

- Tomorrow, I will compare the master values with the previous release and 4.9.x to take a closer look at where these values have increased

**Comment by javiersanchz** [comment]
## Update

I was comparing some more values between `4.10.0-Beta1` and `4.10.0-Alpha3`:


![CPU(%)_grafico1](https://github.com/user-attachments/assets/eeebd7d9-23ba-487f-95ec-04d979d6ed1c)

![FD_grafico1](https://github.com/user-attachments/assets/4b21420a-aa9e-4971-a609-cbb8ea72cbf1)

![RSS(KB)_grafico1](https://github.com/user-attachments/assets/748d5013-b52f-47e9-a006-9bde2d4a3121)

![USS(KB)_grafico1](https://github.com/user-attachments/assets/bc086ff6-06b6-46e6-afb6-c385607b015d)

- Looking at the graphs, it can be observed that `CPU` resources are more or less similar, with some slightly higher peaks.

- `FD` slightly exceeds the previous release version, but there is no significant difference

- The graphs where we can find the most significant differences are in `RSS (KB)` and `USS (KB)`, where there is a peak that notably exceeds the values

**- But what can be observed is that both in the previous release and in this one, the `USS(KB)` and `FD` thresholds were exceeded (in the previous release, this error wouldn't have triggered in the test because the error triggered earlier in `worker-11`), but it suggests that this issue originates from the `4.10.0` branch, not from subsequent changes**

**Comment by javiersanchz** [comment]
## Update

The graphs were created by extracting the resources used over time:

![USS(KB)_grafico](https://github.com/user-attachments/assets/64ce9ee3-dc01-40bb-a17b-87f51e33a0be)

![FD_grafico](https://github.com/user-attachments/assets/84a9ff27-2d8d-4f2c-88d5-48fc25a1a50a)

As we can see, the `FD` exceeds the threshold at all times except at `13:00:00`, due to the manager being restarted

### api.log

```console
2024/11/25 12:59:47 INFO: wazuh 172.31.37.78 "PUT /manager/restart" with parameters {} and body {} done in 0.048s: 200
```

And with it, the cluster as well:

### cluster.log

```console
2024/11/25 12:59:50 DEBUG: [Master] [SendSync] Receiving SendSync request (remoted) from CLUSTER-Workload_benchmarks_metrics_B798_manager_12 (299621)
2024/11/25 12:59:50 INFO: [Master] [Main] SIGNAL [(15)-(SIGTERM)] received. Exit...
2024/11/25 13:01:02 DEBUG: [Cluster] [Main] Removing '/var/ossec/queue/cluster/'.
2024/11/25 13:01:02 DEBUG: [Cluster] [Main] Removed '/var/ossec/queue/cluster/'.
2024/11/25 13:01:02 INFO: [Local Server] [Main] Serving on /var/ossec/queue/cluster/c-internal.sock
2024/11/25 13:01:02 DEBUG: [Local Server] [Keep alive] Calculating.
2024/11/25 13:01:02 DEBUG: [Local Server] [Keep alive] Calculated.
2024/11/25 13:01:02 INFO: [Master] [Main] Serving on ('0.0.0.0', 1516)
2024/11/25 13:01:02 DEBUG: [Master] [Keep alive] Calculating.
2024/11/25 13:01:02 DEBUG: [Master] [Keep alive] Calculated.
2024/11/25 13:01:02 INFO: [Master] [Local integrity] Starting.
```

- The `USS(KB)`, as observed, drops during the restart and then spikes, exceeding the threshold once it starts again.

- On the other hand, the `FD` remains above its threshold, except during the restart

**Comment by javiersanchz** [comment]
## Update

- The root cause of the values exceeding the thresholds after the restart was not identified.

Based on the investigation, the following observations can be made:

- This issue has been occurring in other relesases of `4.10.0`, as seen in the artifacts generated from other releases of this branch, these thresholds are also exceeded, although the test did not show the error (a `worker` failed earlier in those cases https://github.com/wazuh/wazuh/issues/26811).

- On the other hand, it is worth noting that the `FD` exceeds its threshold at all times in the environment except during the restart.

- The `USS (KB)` exceeds the threshold only after the `manager` is restarted and the cluster is brought back up, which suggests the issue may stem from the processes during the cluster startup.

- A comparison was made between potential changes in version `4.9.x` and this branch, but no significant changes were found that could cause this alteration in values.

To rule out any doubts (although it is unlikely to be the cause), in the next environment deployment for the release, increase the value of the `TEST_TIME` parameter to its maximum, `10800` seconds, to evaluate the resources used.

In conclusion, since this issue is scheduled to be closed today, we will wait for the next stage to determine whether it is ultimately a sporadic issue in the test environment or if we need to reopen this issue for further investigation

**Comment by GGP1** [comment]
## Review

Altought the investigation couldn't point out the root cause of the problem, we obtained valuable insights about the resources consumed by the node on each stage. The charts do not match exactly, but that is because one of the tests took longer to complete, if we adjust the `4.10.0-Beta1` values by +50 seconds we get an almost perfect symmetry.

The test failed because of the regression coefficient value which is used to detect high volatility in a node's resources. However, as we have previously seen in https://github.com/wazuh/wazuh/issues/26840, the difference is meaningless and doesn't necessarily represent a performance issue.

LGTM.

**Comment by nico-stefani** [comment]
LGTM
